LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs

```diff
@@ -24,158 +22,150 @@ namespace LoRaWan.NetworkServer.BasicsStation.Processors
     internal class LnsProtocolMessageProcessor : ILnsProtocolMessageProcessor
     {
         private readonly IBasicsStationConfigurationService basicsStationConfigurationService;
+        private readonly WebSocketWriterRegistry<StationEui, string> socketWriterRegistry;
         private readonly ILogger<LnsProtocolMessageProcessor> logger;
 
         public LnsProtocolMessageProcessor(IBasicsStationConfigurationService basicsStationConfigurationService,
+                                           WebSocketWriterRegistry<StationEui, string> socketWriterRegistry,
                                            ILogger<LnsProtocolMessageProcessor> logger)
         {
             this.basicsStationConfigurationService = basicsStationConfigurationService;
+            this.socketWriterRegistry = socketWriterRegistry;
             this.logger = logger;
         }
 
-        public async Task HandleDiscoveryAsync(HttpContext httpContext, CancellationToken token)
+        internal async Task<HttpContext> ProcessIncomingRequestAsync(HttpContext httpContext,
+                                                                     Func<HttpContext, WebSocket, CancellationToken, Task> handler,
+                                                                     CancellationToken cancellationToken)
         {
             if (httpContext is null) throw new ArgumentNullException(nameof(httpContext));
 
-            _ = await ProcessIncomingRequestAsync(httpContext, InternalHandleDiscoveryAsync, token);
-        }
+            if (!httpContext.WebSockets.IsWebSocketRequest)
+            {
+                httpContext.Response.StatusCode = 400;
+                return httpContext;
+            }
 
-        public async Task HandleDataAsync(HttpContext httpContext, CancellationToken token)
-        {
-            if (httpContext is null) throw new ArgumentNullException(nameof(httpContext));
+            using var socket = await httpContext.WebSockets.AcceptWebSocketAsync();
+            this.logger.Log(LogLevel.Debug, $"WebSocket connection from {httpContext.Connection.RemoteIpAddress} established");
+
+            try
+            {
+                await handler(httpContext, socket, cancellationToken);
+                await socket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Goodbye", cancellationToken);
+            }
+            catch (OperationCanceledException ex)
```

Comment:
Not sure if we need to handle this at this stage - we handle this at `WebSocketWriterRegistry.SendAsync` already